### DISCUSSION TITLE
One way video does not work with STUN

### DISCUSSION THREAD

**Discussion by cspiel1** [baresip/baresip#2606] [General]
We found that the NAT pinhole is not opened for one-way video (e.g. early video) if STUN or any Media NAT method is configured.

Following commit adds the function `stream_open_natpinhole()` that had already the exception
```
commit 55ce5afee2e8013bffd14fff49efd75ab13bf180
Author: Christoph Huber <c.huber@commend.com>
Date:   Fri Dec 18 15:05:22 2020 +0100
```

It seems that the function `stream_start_rtcp()` had been the blueprint for @cHuberCoffee commit. This also has this exception. It was added explicitly by one commit from @alfredh.

```
commit aaab7250caf5960af0ca40e3b39f8a140ab07ae4
Author: Alfred E. Heggestad <alfred.heggestad@gmail.com>
Date:   Thu Dec 5 12:19:17 2019 +0100

    stream: dont send RTCP PING if using medianat
```

```c
	if (!strm->mnat) {
		/* Send a dummy RTCP packet to open NAT pinhole */
		err = rtcp_send_app(strm->rtp, "PING", (void *)"PONG", 4);
		if (err) {
			warning("stream: rtcp_send_app failed (%m)\n", err);
			return err;
		}
	}
```

@alfredh, isn't the dummy packet necessary for one-way RTP/RTCP independent from the media NAT configuration?

**Comment by alfredh** [comment]
hi,

could you please try one or more of the following:


- Run a long video call with one-way video, behind NAT. Does the video flow after e.g. 30 minutes ?

- Try to enable RTP/RTCP multiplexing.

- One way video is the same as one way audio, if the audio has no microphone packets, it cannot open the NAT hole.

- Is it possible for you to use ICE instead of STUN ? The STUN protocol for NAT traversal is broken by design.


I think @juha-h has a good setup for this with SER and NatHelper/RTP-Proxy.
Please check with him, but I think his setup is robust and reliable.


Alfred

**Comment by cspiel1** [comment]
Thanks! I'll try to activate my test setup in next days.
@juha-h do you have further information about your NAT setup/config?

**Comment by juha-h** [comment]
I use rtpengine to handle media if one or both parties of a call is behind NAT.  rtpengine is sitting on public IP address and there is thus no need for STUN/TURN/ICE.

But rtpengine is not able to send media packets to a party behind NAT before it has received at least one packet from it.  That packets opens the NAT pinhole and tells rtpengine where (address/port) to send the packets to.

If media is one way, DTLS or ZRTP negotiation is two way and thus rtpengine still receives at least one packet from the party behind NAT.

**Comment by cspiel1** [comment]
Thanks! So this will not help to open the NAT pinhole for one way audio or video.

Then I'll try ICE instead of STUN with the hope that this opens a NAT pinhole.

**Comment by cspiel1** [comment]
Tried to reproduce the issue with `main` HEAD. But the STUN responses from `stun.linphone.org` lead to a call termination. Found that we use a patch internally. Here is a PR: https://github.com/baresip/re/pull/823

**Comment by cspiel1** [comment]
> * Try to enable RTP/RTCP multiplexing.

The first sent RTCP packet opens the pinhole. Then incoming RTP passes. After that e.g. the H.264 decoder has to wait for the first keyframe. This not an acceptable solution.

> * One way video is the same as one way audio, if the audio has no microphone packets, it cannot open the NAT hole.

Exactly.

> * Is it possible for you to use ICE instead of STUN ? The STUN protocol for NAT traversal is broken by design.

What exactly do you mean? How such an account should be configured? Btw: We don't want to use a TURN server.

**Comment by cspiel1** [comment]
I updated the PR. Pinhole opening is only started:
- if mnat is not configured (like before) and
- if STUN is configured

**Comment by alfredh** [comment]
Please read about STUN and ICE and study the RFCs.


Here is a section from RFC 5389:

```
o  Removed the notion that STUN is a complete NAT traversal solution.
      STUN is now a tool that can be used to produce a NAT traversal
      solution.  As a consequence, changed the name of the protocol to
      Session Traversal Utilities for NAT.
```



https://datatracker.ietf.org/doc/html/rfc5389#section-19




STUN as NAT traversal only works with certain types of NAT boxes.
If the NAT is symmetrical, then STUN does not work.
Also please read about this and try to understand it.


What type of NAT box have you been using for this testing?
Have you tried 1 box or multiple different vendors?
Please also share the vendor name and type, if you can.



my modest recommendation would be to remove the `stun.so` module
or at least make it deprecated.


One possible solution would be to make the MediaNAT and Ping
orthogonal, so that either or of them can be enabled with a config option.
For example:

```
rtp_send_ping    yes|no
```

With this option, you can enable NAT ping for any MediaNAT type (if you want)...

**Comment by cspiel1** [comment]
Thanks for the clarification and the proposition of the `rtp_send_ping` setting. I'll open a new PR.

**Comment by cspiel1** [comment]
Here is the closed PR: https://github.com/baresip/baresip/pull/2499
New PR: https://github.com/baresip/baresip/pull/2617
